tests/convergence/test_spatial_methods/test_finite_volume.py

```diff
@@ -0,0 +1,272 @@
+            x = combined_submesh[0].nodes
+            x_edge = combined_submesh[0].edges[1:-1]
+
+            # Define flux and bcs
+            N = pybamm.Vector(x_edge ** 2 * np.cos(x_edge), domain=whole_cell)
+            div_eqn = pybamm.div(N)
+            boundary_conditions = {
+                N.id: {"left": pybamm.Scalar(0), "right": pybamm.Scalar(np.cos(1))}
+            }
+            disc._bcs = boundary_conditions
+            # Define exact solutions
+            # N = x**2 * cos(x) --> dNdx = x*(2cos(x) - xsin(x))
+            div_exact = x * (2 * np.cos(x) - x * np.sin(x))
+
+            # Discretise and evaluate
+            div_eqn_disc = disc.process_symbol(div_eqn)
+            div_approx = div_eqn_disc.evaluate()
+
+            # Return difference between approx and exact
+            return div_approx - div_exact
+
+        # Get errors
+        ns = 10 * 2 ** np.arange(6)
+        errs = {n: get_error(int(n)) for n in ns}
+        # expect quadratic convergence everywhere
+        err_norm = np.array([np.linalg.norm(errs[n], np.inf) for n in ns])
+        rates = np.log2(err_norm[:-1] / err_norm[1:])
+        np.testing.assert_array_less(1.99 * np.ones_like(rates), rates)
+
+    def test_spherical_div_convergence_quadratic(self):
+        # test div( r**2 * sin(r) ) == 4*r*sin(r) - r**2*cos(r)
+        spatial_methods = {"negative particle": pybamm.FiniteVolume}
+
+        # Function for convergence testing
+        def get_error(n):
+            # create mesh and discretisation (single particle)
+            mesh = get_mesh_for_testing(n)
+            disc = pybamm.Discretisation(mesh, spatial_methods)
+            submesh = mesh["negative particle"]
+            r = submesh[0].nodes
+            r_edge = submesh[0].edges[1:-1]
+
+            # Define flux and bcs
+            N = pybamm.Vector(
+                r_edge ** 2 * np.sin(r_edge), domain=["negative particle"]
+            )
+            div_eqn = pybamm.div(N)
+            boundary_conditions = {
+                N.id: {"left": pybamm.Scalar(0), "right": pybamm.Scalar(np.sin(1))}
+            }
+            disc._bcs = boundary_conditions
+            # Define exact solutions
+            # N = r**3 --> div(N) = 5 * r**2
+            div_exact = 4 * r * np.sin(r) + r ** 2 * np.cos(r)
+
+            # Discretise and evaluate
+            div_eqn_disc = disc.process_symbol(div_eqn)
+            div_approx = div_eqn_disc.evaluate()
+
+            # Return difference between approx and exact
+            return div_approx - div_exact
+
+        # Get errors
+        ns = 10 * 2 ** np.arange(6)
+        errs = {n: get_error(int(n)) for n in ns}
+        # expect quadratic convergence everywhere
+        err_norm = np.array([np.linalg.norm(errs[n], np.inf) for n in ns])
+        rates = np.log2(err_norm[:-1] / err_norm[1:])
+        np.testing.assert_array_less(1.99 * np.ones_like(rates), rates)
+
+    def test_spherical_div_convergence_linear(self):
```

Comment:
Maybe @jamiemfoster knows if we lose quadratic convergence for finite volumes on spherical operators. If so maybe the change of variables: u= r * c in the particles might be worth it? (this only works when the diffusion coefficient in the particle is constant though)

Reply:
Is this in reference to the solution of the spherical diffusion equation with a symmetry BC at the core and a state-dependent flux condition at the surface? I.e., the solid transport problem? If so, then an analogous finite volume treatment to the one we were discussing earlier on the week would (I believe) yield second order convergence internally, but would be limited to a global first order rate owing to the extrapolation needed to impose the flux BC at the surface.

I don't think that there are any nuances to do with the spherical operators that would cause you to drop to first order on the internal points. 

Another note about this is that you may well lose perfect conservation in the spherical case. For example, if I write at the internal points that
ci_dot = 1/ri^2 1/delta ( Fi+1/2 - Fi+1/2 )
Then it is not true that Sum Vi ci = constant, where Vi is the volume of the ith shell. To get round this, we can either do a more tailored scheme (e.g. control volumes) or make mapping to slab diffusion as @Scottmar93 suggests. The trouble with the latter is that it gets messy if we state dependent diffusion, which we probably do for most materials.

Reply:
It's a bit simpler than solving a spherical diffusion equation. I'm using a known function of x and seeing how its divergence converges to the real solution (flux is known exactly). But for some forms of the flux I get bad behaviour near `r=0`. If I test that 
```
div( r**2 * sin(r) ) == 4*r*sin(r) - r**2*cos(r)
```
then the inf norm [converges quadratically](https://github.com/pybamm-team/PyBaMM/blob/cd996f357ba0d31e2a9ad4ae3c059cedbd13ae27/tests/convergence/test_spatial_methods/test_finite_volume.py#L138) as I double the mesh size, but if I test 
```
div( r*sin(r) ) == 3*sin(r) + r*cos(r)
```
then the convergence is only [linear](https://github.com/pybamm-team/PyBaMM/blob/cd996f357ba0d31e2a9ad4ae3c059cedbd13ae27/tests/convergence/test_spatial_methods/test_finite_volume.py#L177), with the worst behaviour closest to r=0. Testing 
```
div( sin(r) ) == (2/r)*sin(r) + cos(r)
```
is even worse, the numerical solution tends to 4 as r-->0, instead of tending to 3 as you'd expect, independently of grid size

Good point on conservation, we'll have to keep an eye on that